game/world/objects/npc.cpp

```diff
@@ -3448,26 +3461,38 @@ bool Npc::tickCast(uint64_t dt) {
   if(bodyStateMasked()!=BS_CASTING)
     return true;
 
-  auto& spl = owner.script().spellDesc(active->spellId());
-  int32_t castLvl = int(castLevel)-int(CS_Invest_0);
   if(owner.tickCount()<castNextTime)
     return true;
 
-  if(castLvl<=15)
-    castLevel = CastState(castLevel+1);
+  ++manaInvested;
+  const SpellCode code = SpellCode(owner.script().invokeMana(*this,currentTarget,manaInvested));
+
+  if(owner.version().game==1) {
+    changeAttribute(ATR_MANA,-1,false);
+    if(!isPlayer() && code!=SpellCode::SPL_SENDCAST)
+      assert(attribute(ATR_MANA)>0);
+    }
 
-  int32_t mana = attribute(ATR_MANA);
-  const SpellCode code = SpellCode(owner.script().invokeMana(*this,currentTarget,*active));
-  mana = std::max(mana-attribute(ATR_MANA),0);
+  if(!isPlayer() && aiExpectedInvest<=manaInvested) {
+    endCastSpell();
+    return true;
+    }
+
+  int32_t castLvl = int(castLevel)-int(CS_Invest_0);
 
   switch(code) {
     case SpellCode::SPL_NEXTLEVEL: {
+      if(castLvl<=15)
+        castLevel = CastState(castLevel+1);
+      }
+    case SpellCode::SPL_RECEIVEINVEST:
+    case SpellCode::SPL_STATUS_CANINVEST_NO_MANADEC: {
+      auto& spl = owner.script().spellDesc(active->spellId());
       visual.setMagicWeaponKey(owner,SpellFxKey::Invest,castLvl+1);
```

Comment:
Tested today:
apparently `setMagicWeaponKey` must go into `SPL_NEXTLEVEL`, otherwise any spell in G2 has broken sound/visual